keps/sig-storage/4876-mutable-csinode-allocatable/README.md

```diff
@@ -0,0 +1,843 @@
+compatibility.
+
+#### Volume Plugin Manager
+
+A new goroutine will be started in VolumePluginMgr’s [Run()](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/plugins.go#L953) func if the `AllocatableUpdateInterval` is set to a nonzero value. This goroutine will periodically trigger updates to the `CSINode` object based on the specified interval:
+
+```golang
+func (pm *VolumePluginMgr) Run(stopCh <-chan struct{}) {
+    if pm.csiNodeUpdateInterval > 0 {
+        go wait.Until(pm.updateCSINodeInfo, pm.csiNodeUpdateInterval, stopCh)
+    }
+}
+```
+
+#### NodeInfoManager Interface Extension
+
+The existing [NodeInfoManager](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/csi/nodeinfomanager/nodeinfomanager.go#L76) interface will be extended to include a new method for updating the `CSINode` object:
+
+```golang
+// Interface implements an interface for managing labels of a node
+type Interface interface {
+    CreateCSINode() (*storagev1.CSINode, error)
+    ...
+    // UpdateCSINode updates the CSINode object
+    UpdateCSINode() error
+}
+```
+
+#### CSINode Update Mechanism
+
+- If `AllocatableUpdateInterval` is set in the `CSIDriver` object, Kubernetes will periodically call the CSI driver’s `NodeGetInfo` RPC endpoint at the defined interval, and update the `CSINode` object with the latest `Allocatable.Count`.
+
+- If `AllocatableUpdateInterval` is not set, updates will only occur in response to a failure during volume attachment if the feature flag is set. For example, if the CSI driver returns a `ResourceExhausted` error, Kubernetes will immediately update the `CSINode` object so that the value more accurately reflects the actual state of the world.
+
+The existing `NodeInfoManager` interface will be used to perform the actual update for consistency with established patterns.
+
+#### Pod Construction Changes
+
+To address race conditions where the scheduler assigns stateful pods to nodes with insufficient capacity, Kubelet's pod construction process during [WaitForAttachAndMount](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/volume_manager.go#L393) will now handle `ResourceExhausted` errors returned by CSI drivers during the `ControllerPublishVolume` RPC.
+
+The `ResourceExhausted` error is directly reported on the `VolumeAttachment` object associated with the relevant attachment.
+
+```golang
+if err := kl.volumeManager.WaitForAttachAndMount(pod); err != nil {
+    if isResourceExhaustedError(err) {
+        // Update CSINode
+        // Generate event for affected pod
+    } else {
+        // Existing error handling
+    }
+}
+```
+
+This change ensures that when a pod fails to be constructed due to insufficient volume attachment capacity, that both:
+
+1. The `CSINode` object is promptly updated to reflect the actual available capacity, improving future scheduling decisions.
```

Comment:
I would consider some sort of backoff mechanism

Reply:
That makes sense, incorporated this feedback in latest rev.